File: oxidize-python/oxidize_python/internal/generate/paged_run.py
"""PagedAttention generation path mirroring oxidize-golang/internal/generate/paged_run.go."""

from __future__ import annotations

import time
from typing import TYPE_CHECKING, Protocol

if TYPE_CHECKING:
    from oxidize_python.internal.generate.runtime import RunConfig

from oxidize_python.core.model.loader import LoaderConfig, load_gguf_model_from_path
from oxidize_python.core.model.model import Model, Session, Token
from oxidize_python.core.model.sampling import greedy
from oxidize_python.core.paged.paged import (
    Scheduler,
    SchedulerV2,
    SchedulerV2Config,
    SequenceStatus,
    default_scheduler_config,
    default_scheduler_v2_config,
)
from oxidize_python.core.tokenizer import from_gguf_metadata
from oxidize_python.core.tokenizer.bpe import BpeTokenizer
from oxidize_python.core.tokenizer.tokenizer import EncodeOptions, SpecialTokens
from oxidize_python.internal.gguf.parse import load_file


def _load_tokenizer(path: str):
    try:
        gguf = load_file(path)
        meta = {k: v.string for k, v in gguf.metadata.items() if v.string}
        return from_gguf_metadata(meta)
    except Exception:
        return BpeTokenizer([], [], SpecialTokens(bos=1, eos=2))


class _Stdout(Protocol):
    def write(self, s: str, /) -> object: ...


class PagedGenerateRuntime:
    def __init__(self, model_path: str) -> None:
        loaded = load_gguf_model_from_path(model_path, LoaderConfig())
        self.sched = Scheduler(default_scheduler_config())
        self.mdl: Model = loaded
        self.sessions: dict[int, Session] = {}
        self.prompts: dict[int, list[int]] = {}
        self.max_tok: dict[int, int] = {}
        self.generated: dict[int, int] = {}

    def enqueue(self, prompt_tokens: list[int], max_new: int) -> int:
        req = self.sched.add_request(prompt_tokens, max_new)
        rid = req.id
        self.prompts[rid] = list(prompt_tokens)
        self.max_tok[rid] = max_new
        self.generated[rid] = 0
        self.sessions[rid] = Session()
        return rid

    def step(self) -> dict[int, Token]:
        active = self.sched.step()
        if not active:
            return {}
        out: dict[int, Token] = {}
        for req in active:
            rid = req.id
            sess = self.sessions[rid]
            tokens = self.prompts[rid]
            if self.generated[rid] == 0 and len(tokens) > 1:
                batch = [Token(t) for t in tokens[:-1]]
                self.mdl.forward(batch, sess)
            last = tokens[-1]
            logits = self.mdl.forward([Token(last)], sess)
            next_tok = greedy(logits)
            out[rid] = next_tok
            self.prompts[rid] = [*tokens, int(next_tok)]
            self.generated[rid] += 1
            if self.generated[rid] >= self.max_tok[rid] or next_tok == 2:
                self.sched.finish(rid)
                self.sessions.pop(rid, None)
                self.prompts.pop(rid, None)
                self.max_tok.pop(rid, None)
                self.generated.pop(rid, None)
        return out


class PagedGenerateRuntimeV2:
    """Budgeted three-phase paged runtime mirroring
    oxidize-golang/internal/server/paged_runtime.go::PagedRuntimeV2.

    It enforces a token budget per step, supports prefill chunking and prefix
    caching, and builds an InputBatch each step so multiple sequences can be
    processed together. The model forward is still driven per-sequence (one
    session each) because the pure-Python model backend does not yet expose a
    fused multi-sequence kernel, but scheduling, batching metadata, and block
    management mirror the Rust scheduler.
    """

    def __init__(
        self,
        model_path: str,
        config: SchedulerV2Config | None = None,
        total_blocks: int = 1024,
        block_size: int = 16,
    ) -> None:
        loaded = load_gguf_model_from_path(model_path, LoaderConfig())
        self.mdl: Model = loaded
        self.sched = SchedulerV2(
            config or default_scheduler_v2_config(), total_blocks, block_size
        )
        self.sessions: dict[int, Session] = {}
        self.model_prefilled: dict[int, int] = {}

    def enqueue(
        self,
        prompt_tokens: list[int],
        max_new: int,
        stop_token: int = 2,
        has_stop: bool = True,
    ) -> int:
        seq_id = self.sched.add_request(prompt_tokens, max_new, stop_token, has_stop)
        self.sessions[seq_id] = Session()
        self.model_prefilled[seq_id] = 0
        return seq_id

    def step(self) -> dict[int, Token]:
        res = self.sched.step()
        if not res.scheduled_seq_ids:
            return {}
        batch = self.sched.build_input_batch(res)

        sampled: dict[int, int] = {}
        out: dict[int, Token] = {}
        for i, seq_id in enumerate(batch.seq_ids):
            sess = self.sessions.get(seq_id)
            if sess is None:
                sess = Session()
                self.sessions[seq_id] = sess
            toks = [Token(t) for t in batch.token_ids[i]]
            if not toks:
                continue
            logits = self.mdl.forward(toks, sess)
            # Only sequences finishing their prefill (or decoding) produce a
            # sampled token this step. For a prefill chunk that does not yet
            # reach the end of the prompt, we keep accumulating context and skip
            # sampling so we do not emit mid-prompt tokens.
            if batch.is_prefill[i]:
                seq = self.sched.get_sequence(seq_id)
                if seq is None or seq.remaining_prefill_tokens() > 0:
                    continue
            next_tok = greedy(logits)
            sampled[seq_id] = int(next_tok)
            out[seq_id] = next_tok

        self.sched.postprocess_step(sampled)
        # Reap fully finished sequences' sessions.
        for seq_id in list(out.keys()):
            seq = self.sched.get_sequence(seq_id)
            if seq is not None and seq.status == SequenceStatus.FINISHED:
                self.sessions.pop(seq_id, None)
                self.model_prefilled.pop(seq_id, None)
        return out

    def stats(self) -> tuple[int, int, int]:
        return (
            self.sched.waiting_count(),
            self.sched.running_count(),
            self.sched.pool.free_count(),
        )


def run_paged_from_gguf(cfg: "RunConfig", stdout: _Stdout) -> None:
    path = cfg.model_path.strip()
    prompt = cfg.prompt.strip()
    if not path:
        raise ValueError("generate: empty model path")
    if not prompt:
        return

    tok = _load_tokenizer(path)
    prompt_tokens = tok.encode(prompt, EncodeOptions()) or [Token(1)]
    int_prompt = [int(t) for t in prompt_tokens]

    runtime = PagedGenerateRuntime(path)
    req_id = runtime.enqueue(int_prompt, cfg.max_new_tokens)

    start = time.monotonic()
    generated = 0
    stop = cfg.stop_token or Token(2)
    while generated < cfg.max_new_tokens:
        tokens = runtime.step()
        tok_out = tokens.get(req_id)
        if tok_out is None:
            if not tokens:
                break
            continue
        try:
            piece = tok.decode([tok_out])
        except Exception:
            piece = f"<{tok_out}>"
        stdout.write(piece)
        generated += 1
        if tok_out == stop or tok_out == 2:
            break

    elapsed = time.monotonic() - start
    if elapsed > 0 and generated > 0:
        speed = generated / elapsed
        stdout.write(f"\ngeneration stats: tokens={generated} speed={speed:.2f} tok/s (paged)\n")

File: oxidize-golang/internal/generate/paged_run.go
package generate

import (
	"context"
	"fmt"
	"io"
	"strings"
	"time"

	"github.com/Zapdev-labs/oxidize/golang/core/model"
	"github.com/Zapdev-labs/oxidize/golang/core/paged"
	"github.com/Zapdev-labs/oxidize/golang/core/tokenizer"
)

type pagedGenerateRuntime struct {
	sched     *paged.Scheduler
	mdl       model.Model
	sessions  map[int]*model.Session
	prompts   map[int][]int
	maxTok    map[int]int
	generated map[int]int
}

func newPagedGenerateRuntime(modelPath string, schedCfg paged.SchedulerConfig) (*pagedGenerateRuntime, error) {
	m, err := LoadModelFromPath(modelPath, LoaderConfig{AllowFallback: true})
	if err != nil {
		return nil, err
	}
	return &pagedGenerateRuntime{
		sched:     paged.NewScheduler(schedCfg),
		mdl:       m.Model,
		sessions:  make(map[int]*model.Session),
		prompts:   make(map[int][]int),
		maxTok:    make(map[int]int),
		generated: make(map[int]int),
	}, nil
}

func (r *pagedGenerateRuntime) enqueue(promptTokens []int, maxNew int) (int, error) {
	req, err := r.sched.AddRequest(promptTokens, maxNew)
	if err != nil {
		return 0, err
	}
	id := req.ID
	r.prompts[id] = append([]int(nil), promptTokens...)
	r.maxTok[id] = maxNew
	r.generated[id] = 0
	r.sessions[id] = model.NewSession()
	return id, nil
}

func (r *pagedGenerateRuntime) step(ctx context.Context) (map[int]model.Token, error) {
	if err := ctx.Err(); err != nil {
		return nil, err
	}
	active, err := r.sched.Step()
	if err != nil {
		return nil, err
	}
	out := make(map[int]model.Token, len(active))
	for _, req := range active {
		id := req.ID
		sess := r.sessions[id]
		tokens := r.prompts[id]
		if r.generated[id] == 0 && len(tokens) > 1 {
			batch := make([]model.Token, len(tokens)-1)
			for i, t := range tokens[:len(tokens)-1] {
				batch[i] = model.Token(t)
			}
			if _, err := r.mdl.Forward(batch, sess); err != nil {
				return nil, err
			}
		}
		last := tokens[len(tokens)-1]
		logits, err := r.mdl.Forward([]model.Token{model.Token(last)}, sess)
		if err != nil {
			return nil, err
		}
		next, err := model.Greedy(logits)
		if err != nil {
			return nil, err
		}
		out[id] = next
		r.prompts[id] = append(tokens, int(next))
		r.generated[id]++
		if r.generated[id] >= r.maxTok[id] || next == 2 {
			r.sched.Finish(id)
			delete(r.sessions, id)
			delete(r.prompts, id)
			delete(r.maxTok, id)
			delete(r.generated, id)
		}
	}
	return out, nil
}

// RunPagedFromGGUF uses the vLLM-style paged scheduler for generation.
func RunPagedFromGGUF(ctx context.Context, cfg RunConfig, stdout io.Writer) error {
	if strings.TrimSpace(cfg.ModelPath) == "" {
		return fmt.Errorf("generate: empty model path")
	}
	if strings.TrimSpace(cfg.Prompt) == "" {
		return nil
	}

	tok, err := loadTokenizer(cfg.ModelPath, cfg.TokenizerModel)
	if err != nil {
		return err
	}
	promptTokens, err := tok.Encode(cfg.Prompt, tokenizer.EncodeOptions{})
	if err != nil {
		return fmt.Errorf("encode: %w", err)
	}
	if len(promptTokens) == 0 {
		promptTokens = []model.Token{1}
	}
	intPrompt := make([]int, len(promptTokens))
	for i, t := range promptTokens {
		intPrompt[i] = int(t)
	}

	runtime, err := newPagedGenerateRuntime(cfg.ModelPath, paged.DefaultSchedulerConfig())
	if err != nil {
		return err
	}
	reqID, err := runtime.enqueue(intPrompt, cfg.MaxNewTokens)
	if err != nil {
		return err
	}

	start := time.Now()
	generated := 0
	for generated < cfg.MaxNewTokens {
		tokens, err := runtime.step(ctx)
		if err != nil {
			return err
		}
		tokOut, ok := tokens[reqID]
		if !ok {
			if len(tokens) == 0 {
				break
			}
			continue
		}
		piece, err := tok.Decode([]model.Token{tokOut})
		if err != nil {
			piece = fmt.Sprintf("<%d>", tokOut)
		}
		if _, err := io.WriteString(stdout, piece); err != nil {
			return err
		}
		generated++
		if tokOut == cfg.StopToken || tokOut == 2 {
			break
		}
	}
	elapsed := time.Since(start).Seconds()
	if elapsed > 0 && generated > 0 {
		_, _ = fmt.Fprintf(stdout, "\ngeneration stats: tokens=%d speed=%.2f tok/s (paged)\n", generated, float64(generated)/elapsed)
	}
	return nil
}
